### PR TITLE
statistics: use another way to merge topn

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: ref #50761

Problem Summary:

### What is changed and how it works?

We use the property that items inside both the TopN and the histogram are ordered to speed up the process.

And use a heuristic cutting:
- We record the avg num per distinct value of each histogram `notNullCount / ndv_in_hist`.
- Then for a topn item, we can use the mentioned property to know the TopNs that contain the item.
- Then we can easily know in which partition the item occurs in the histogram.
- The maximum possible occurrence of this item is `the sum occurrence of the affected TopNs + the avg num per distinct value of each affected histogram`
- If the maximum possible occurrence is still smaller than the currently maintained smallest global TopN item, we don't need to insert this one into the heap and test it.

In this way, the speed is improved while the CPU is saved.
img_v2_3e92f289-37de-4fd4-9b33-2a35f25c9d5g
The CPU usage. Previous VS This pull.

```
Running tool: /DATA/disk4/yiding/go/bin/go test -benchmem -run=^$ -tags intest,deadlock -bench ^BenchmarkMergePartTopN2GlobalTopNWithHists$ github.com/pingcap/tidb/pkg/statistics/handle/globalstats

goos: linux
goarch: amd64
pkg: github.com/pingcap/tidb/pkg/statistics/handle/globalstats
cpu: Intel(R) Xeon(R) Gold 6240 CPU @ 2.60GHz
BenchmarkMergePartTopN2GlobalTopNWithHists/Size100-72         	     159	   7249785 ns/op	   34714 B/op	     113 allocs/op
BenchmarkMergePartTopN2GlobalTopNWithHists/Size1000-72        	       9	 117386869 ns/op	  100136 B/op	    1013 allocs/op
BenchmarkMergePartTopN2GlobalTopNWithHists/Size2000-72        	       4	 272247218 ns/op	  173416 B/op	    2013 allocs/op
BenchmarkMergePartTopN2GlobalTopNWithHists/Size5000-72        	       2	 787035728 ns/op	  393256 B/op	    5013 allocs/op
BenchmarkMergePartTopN2GlobalTopNWithHists/Size10000-72       	       1	1919716271 ns/op	  759656 B/op	   10013 allocs/op
PASS
ok  	github.com/pingcap/tidb/pkg/statistics/handle/globalstats	17.839s

Running tool: /DATA/disk4/yiding/go/bin/go test -benchmem -run=^$ -tags intest,deadlock -bench ^BenchmarkMergePartTopN2GlobalTopNWithHists$ github.com/pingcap/tidb/pkg/statistics/handle/globalstats

goos: linux
goarch: amd64
pkg: github.com/pingcap/tidb/pkg/statistics/handle/globalstats
cpu: Intel(R) Xeon(R) Gold 6240 CPU @ 2.60GHz
BenchmarkMergePartTopN2GlobalTopNWithHists/Size100-72         	     122	  10071203 ns/op	  177982 B/op	      30 allocs/op
BenchmarkMergePartTopN2GlobalTopNWithHists/Size1000-72        	       6	 186292872 ns/op	  176418 B/op	      28 allocs/op
BenchmarkMergePartTopN2GlobalTopNWithHists/Size2000-72        	       3	 398591599 ns/op	  178269 B/op	      31 allocs/op
BenchmarkMergePartTopN2GlobalTopNWithHists/Size5000-72        	       1	1436609512 ns/op	  179432 B/op	      33 allocs/op
BenchmarkMergePartTopN2GlobalTopNWithHists/Size10000-72       	       1	4363627038 ns/op	  177336 B/op	      31 allocs/op
PASS
ok  	github.com/pingcap/tidb/pkg/statistics/handle/globalstats	18.587s
```

```
n=100,            7249785/10071203=72%
n=1000         117386869/186292872=63%
n=2000         272247218/398591599=68%
n=5000        787035728/1436609512=55%
n=10000      1919716271/4363627038=44%
```

The mem usage is also in an acceptable range.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/statistics/handle/globalstats/global_stats.go [hunk 1]
```diff
@@ -151,7 +151,7 @@ func MergePartitionStats2GlobalStatsByTableID(
 // These values originally came from the analyzeOptionDefault structure in the planner/core/planbuilder.go file.
 var analyzeOptionDefault = map[ast.AnalyzeOptionType]uint64{
 	ast.AnalyzeOptNumBuckets: 256,
-	ast.AnalyzeOptNumTopN:    20,
```
Comment (winoros) [inline pkg/statistics/handle/globalstats/global_stats.go:154]: 20 is a wrong value compared with the current non-partitioned table. But, yes, the previous merging operation is too slow. So I want to change it together with this pr.

**Comment by winoros** [comment]
And some codes like the in-place updates for the heap are optimized for the memory.

File: pkg/statistics/handle/globalstats/global_stats.go (hunk 1 above)
Comment (winoros) [reply pkg/statistics/handle/globalstats/global_stats.go]: Undo it and change it later for smaller diff.

File: pkg/statistics/handle/globalstats/topn.go [hunk 2]
```diff
@@ -138,7 +262,173 @@ func MergeGlobalStatsTopNByConcurrency(
 //     just delete some values when we merge the global-level topN.
```
Comment (elsa0520) [inline pkg/statistics/handle/globalstats/topn.go:262]: Could you please add the algorithm logic in the comments ~
(Especially the pruning algorithm)

File: pkg/statistics/handle/globalstats/topn.go [hunk 3]
```diff
@@ -138,7 +262,173 @@ func MergeGlobalStatsTopNByConcurrency(
...
+		iter.finish()
+	}
+	statistics.SortTopnMeta(finalTopNs)
+	statistics.SortTopnMeta(remainedTopNs)
+	var globalTopN statistics.TopN
+	globalTopN.TopN = finalTopNs
+	globalTopN.Sort()
+	return &globalTopN, remainedTopNs, hists, nil
+}
+
```
Comment (elsa0520) [inline pkg/statistics/handle/globalstats/topn.go:428]: Mark as deprecated

File: pkg/statistics/handle/globalstats/global_stats_test.go [hunk 4]
```diff
@@ -921,6 +921,7 @@ func TestGlobalIndexStatistics(t *testing.T) {
 }
 
 func TestIssues24349(t *testing.T) {
+	t.Skip()
```
Comment (elsa0520) [inline pkg/statistics/handle/globalstats/global_stats_test.go:924]: If we keep version 1 analyze, it is best to run version 1 test. Or just delete the case directly

File: pkg/statistics/cmsketch_util.go [hunk 5]
```diff
@@ -72,3 +72,24 @@ func topNMetaToDatum(val TopNMeta,
 	}
 	return dat, err
 }
+
+// TopNMetaValToDatum transforms the value of the TopN item to the datum.
+func TopNMetaValToDatum(val []byte, tp byte, isIndex bool, loc *time.Location) (d types.Datum, err error) {
+	if isIndex {
+		d.SetBytes(val)
+	} else {
+		var err error
```
Comment (hi-rustin) [inline pkg/statistics/cmsketch_util.go:81]: I guess we don't need to redeclare it here.

File: pkg/statistics/cmsketch_util.go [hunk 6]
```diff
@@ -72,3 +72,24 @@ func topNMetaToDatum(val TopNMeta,
...
+		if types.IsTypeTime(tp) {
+			// Handle date time values specially since they are encoded to int and we'll get int values if using DecodeOne.
+			_, d, err = codec.DecodeAsDateTime(val, tp, loc)
+		} else if types.IsTypeFloat(tp) {
+			_, d, err = codec.DecodeAsFloat32(val, tp)
+		} else {
+			_, d, err = codec.DecodeOne(val)
+		}
+		if err != nil {
```
Comment (hi-rustin) [inline pkg/statistics/cmsketch_util.go:90]: I guess we can also skip this check.

File: pkg/statistics/handle/globalstats/global_stats_test.go (hunk 4 above)
Comment (hi-rustin) [reply pkg/statistics/handle/globalstats/global_stats_test.go:924]: +1. Maybe we should update it or set the version to v1.

File: pkg/statistics/handle/globalstats/topn.go [hunk 7]
```diff
@@ -121,6 +130,121 @@ func MergeGlobalStatsTopNByConcurrency(
 	return globalTopN, popedTopn, wrapper.AllHg, nil
 }
 
+type histIter struct {
+	hist             *statistics.Histogram
+	totalSubstracted int64
+	curBucketPos     int
```
Comment (hi-rustin) [inline pkg/statistics/handle/globalstats/topn.go:136]: Better to add comments to these fields.

File: pkg/statistics/handle/globalstats/topn.go [hunk 8]
```diff
@@ -138,7 +262,173 @@ func MergeGlobalStatsTopNByConcurrency(
...
+	}
+	maxPossibleAdded := make([]int64, len(hists))
+	for i, hist := range hists {
+		curMax := int64(hist.NotNullCount() / float64(hist.NDV))
+		for _, bkt := range hist.Buckets {
+			curMax = max(curMax, bkt.Repeat)
+		}
+		maxPossibleAdded[i] = curMax
+	}
+	if mergingHeap.Len() == 0 {
```
Comment (hi-rustin) [inline pkg/statistics/handle/globalstats/topn.go:292]: It seems we can check this before we calculate maxPossibleAdded.

File: pkg/statistics/handle/globalstats/topn.go [hunk 9]
```diff
@@ -138,7 +262,173 @@ func MergeGlobalStatsTopNByConcurrency(
...
+				return nil, nil, nil, err
+			}
+			continue
+		}
+		// The cmp result cannot be 1 because the value is strictly increasing.
+		// Here is cmp == 0.
+		cur.item.Count += headTopN.Count
+		cur.affectedTopNs.Set(uint(head.idx))
+	}
+	{
```
Comment (hi-rustin) [inline pkg/statistics/handle/globalstats/topn.go:411]: Why do we need a separate block here?

File: pkg/statistics/handle/globalstats/topn.go [hunk 10]
```diff
@@ -138,7 +262,173 @@ func MergeGlobalStatsTopNByConcurrency(
...
+		affectedTopNs: bitset.New(uint(len(hists))),
+	}
+	step := int64(0)
+	histRemoveCnt := int64(0)
+	var finalTopNs topNMeataHeap = make([]statistics.TopNMeta, 0, n+1)
+	remainedTopNs := make([]statistics.TopNMeta, 0, n)
+	skipCount := 0
+	affectedHist := make([]int, 0, len(hists))
+	firstTime := true
+	checkTheCurAndMoveForward := func(nextVal *statistics.TopNMeta, position uint) error {
```
Comment (hi-rustin) [inline pkg/statistics/handle/globalstats/topn.go:313]: Maybe better to move this function out. Because it is quite long.

**Review by 0xPoe — COMMENTED** [review]
Just a quick run through the code, I'm still trying to understand the business logic of it. I'll look at it a couple more times soon to try and understand it.

File: pkg/statistics/handle/globalstats/topn.go (hunk 10 above)
Comment (winoros) [reply pkg/statistics/handle/globalstats/topn.go:313]: If so, I prefer just inlining it, or adding another wrapper or helper.

File: pkg/statistics/handle/globalstats/topn.go (hunk 9 above)
Comment (winoros) [reply pkg/statistics/handle/globalstats/topn.go:411]: The block is added before i extract that func. It's a longer one compared at that time.

File: pkg/statistics/handle/globalstats/topn.go (hunk 7 above)
Comment (winoros) [reply pkg/statistics/handle/globalstats/topn.go:140]: added

File: pkg/statistics/handle/globalstats/topn.go (hunk 2 above)
Comment (winoros) [reply pkg/statistics/handle/globalstats/topn.go:266]: added

File: pkg/statistics/handle/globalstats/global_stats_test.go (hunk 4 above)
Comment (winoros) [reply pkg/statistics/handle/globalstats/global_stats_test.go:924]: There might be another problem, but I'm still trying not to skip or use v1 to test it.

File: pkg/statistics/handle/globalstats/topn.go [hunk 11]
```diff
@@ -136,9 +264,180 @@ func MergeGlobalStatsTopNByConcurrency(
...
+		if err != nil {
+			return err
+		}
+		affectedHist = affectedHist[:0]
+		// The following codes might access the NextClear loop twice. Record it here for saving CPU.
+		for histPos, found := cur.affectedTopNs.NextClear(0); found; histPos, found = cur.affectedTopNs.NextClear(histPos + 1) {
+			affectedHist = append(affectedHist, int(histPos))
+		}
+		// Hacking skip.
+		if uint32(len(finalTopNs)) >= n {
```
Comment (hi-rustin) [inline pkg/statistics/handle/globalstats/topn.go:335]: I am not sure which one we prefer. `len(finalTopNs)` vs. `finalTopNs.Len()`. But I guess it doesn't matter.

File: pkg/statistics/handle/globalstats/topn.go (hunk 10 above)
Comment (hi-rustin) [reply pkg/statistics/handle/globalstats/topn.go:322]: Another wrapper sounds good to me.

**Review by 0xPoe — APPROVED** [review]
Looks good to me. But please move checkTheCurAndMoveForward out as a wrapper and add more comments. Thank you for working on this. Approve in advance.

**Review by elsa0520 — APPROVED** [review]
LGTM

**Comment by 0xPoe** [comment]
/hold

I would like to take another look.

File: pkg/statistics/handle/globalstats/topn.go [hunk 12]
```diff
@@ -177,7 +509,7 @@ func MergePartTopN2GlobalTopN(
 					return nil, nil, nil, err
 				}
 
-				if (j == i && version >= 2) || topNs[j].FindTopN(val.Encoded) != -1 {
+				if topNs[j].FindTopN(val.Encoded) != -1 {
```
Comment (0xPoe) [inline pkg/statistics/handle/globalstats/topn.go:512]: May I ask why we also remove the `j==i` check here?

File: pkg/statistics/handle/globalstats/topn.go [hunk 13]
```diff
@@ -136,9 +388,89 @@ func MergeGlobalStatsTopNByConcurrency(
 //     but is not placed to global-level TopN. We should put them back to histogram latter.
 //  3. `[]*Histogram` are the partition-level histograms which
 //     just delete some values when we merge the global-level topN.
+//
+// The function use the merging operation based on the property that the topn items are ordered.
+// And it does a heuristic short-cutting:
+// - for newly seen merged topn item, it's total occurrence is the sum in the topn and the sum in the hist.
```
Comment (0xPoe) [inline pkg/statistics/handle/globalstats/topn.go:394]:
```suggestion
// The function use the merging operation based on the property that the topN items are ordered.
// And it does a heuristic short-cutting:
// - for newly seen merged topN item, it's total occurrence is the sum in the topN and the sum in the hist.
```

File: pkg/statistics/handle/globalstats/topn.go [hunk 14]
```diff
@@ -136,9 +388,89 @@ func MergeGlobalStatsTopNByConcurrency(
...
+// - its occurrence in each hist is calculated as row_num_hist / ndv_in_hist or 0 or the bucket bound.
+// - so we can calculate its maximum possible added value to max(row_num_hist / ndv_in_hist, bucket bound of each bucket) for each hist + sum_in_topn
+// - we don't need to actually check it histogram by histogram if the maximum possible occurrence is still smaller than smallest maintained topn.
```
Comment (0xPoe) [inline pkg/statistics/handle/globalstats/topn.go:397]:
```suggestion
// - so we can calculate its maximum possible added value to max(row_num_hist / ndv_in_hist, bucket bound of each bucket) for each hist + sum_in_topn
// - we don't need to actually check it histogram by histogram if the maximum possible occurrence is still smaller than the smallest maintained topN.
```

File: pkg/statistics/handle/globalstats/topn.go [hunk 15]
```diff
@@ -121,6 +130,249 @@ func MergeGlobalStatsTopNByConcurrency(
...
+
+func (h *topNMeataHeap) Pop() any {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+type topNMerger struct {
```
Comment (0xPoe) [inline pkg/statistics/handle/globalstats/topn.go:252]: Do you mind if we add more detailed tests for this merger to cover the heuristic rules and edge cases? Also, the real test would be a good entry point to understand how to use it and how it works.

**Review by 0xPoe — COMMENTED** [review]
I will spend more time on topNMerger as I am still trying to understand it.

Thanks!

File: pkg/statistics/handle/globalstats/topn.go [hunk 16]
```diff
@@ -136,9 +388,89 @@ func MergeGlobalStatsTopNByConcurrency(
...
+		if err := killer.HandleSignal(); err != nil {
+			return nil, nil, nil, err
+		}
+		iter.finish()
+	}
+	statistics.SortTopnMeta(merger.finalTopNs)
+	statistics.SortTopnMeta(merger.remainedTopNs)
+	var globalTopN statistics.TopN
+	globalTopN.TopN = merger.finalTopNs
+	globalTopN.Sort()
```
Comment (0xPoe) [inline pkg/statistics/handle/globalstats/topn.go:467]: It seems this is a double sort?

File: pkg/statistics/handle/globalstats/topn.go [hunk 17]
```diff
@@ -121,6 +130,249 @@ func MergeGlobalStatsTopNByConcurrency(
...
+	finalTopNs          topNMeataHeap
+	remainedTopNs       []statistics.TopNMeta
+	cur                 struct {
+		affectedTopNs *bitset.BitSet
+		item          statistics.TopNMeta
+	}
+
+	n       int
+	isIndex bool
+	tp      byte
```
Comment (0xPoe) [inline pkg/statistics/handle/globalstats/topn.go:268]: Could you please add some detailed comments for these fields? Thank you!

**Review by 0xPoe — COMMENTED** [review]
Thanks!

**Review by 0xPoe — COMMENTED** [review]
Thanks!

**Comment by winoros** [comment]
This pr optimized the original algorithm, but the time complexity is still large because we keep the core calculation unchanged.
Now that we have a new issue to track the performance. I close this pr.
